driver/driver.c

```diff
@@ -71,7 +95,184 @@ static const struct euicc_driver *_driver_http = NULL;
+    return strdup(dirname(buf));
+}
+#elif defined(_WIN32)
+static char *get_origin() {
+    char *buf = (char *)calloc(MAX_PATH, sizeof(char));
+    if (GetModuleFileNameA(NULL, buf, MAX_PATH) != 0) {
+        return buf;
+    } else {
+        fprintf(stderr, "Failed to get $ORIGIN.\n");
+        return NULL;
+    }
+}
+#elif defined(__APPLE__) && defined(__MACH__)
+static char *get_origin() {
+    char *buf = (char *)calloc(PATH_MAX, sizeof(char));
+    uint32_t size = PATH_MAX;
+    if (_NSGetExecutablePath(buf, &size) == 0) {
+        return buf;
+    } else {
+        return NULL;
+    }
+}
+#endif
+
+#if defined(__linux__)
+static const char *get_runpath() {
+    void *handle = dlopen(NULL, RTLD_NOW);
+    if (handle == NULL) {
+        return NULL;
+    }
+    struct link_map *linkmap;
+    if (dlinfo(handle, RTLD_DI_LINKMAP, &linkmap) == -1) {
+        return NULL;
+    }
+    const ElfW(Dyn) *dyn = linkmap->l_ld;
+    const ElfW(Dyn) *rpath = NULL;
+    const ElfW(Dyn) *runpath = NULL;
+    const char *strtab = NULL;
+    for (; dyn->d_tag != DT_NULL; ++dyn) {
+        if (dyn->d_tag == DT_RPATH) {
+            rpath = dyn;
+        } else if (dyn->d_tag == DT_RUNPATH) {
+            runpath = dyn;
+        } else if (dyn->d_tag == DT_STRTAB) {
+            strtab = (const char *)dyn->d_un.d_val;
+        }
+    }
+    if (runpath) {
+        return strtab + runpath->d_un.d_val;
+    } else {
+        return NULL;
+    }
+}
+
+#else
+// FIXME: I still didn't find the way to determine RUNPATH on non-Linux.
+//        So use $ORIGIN as a temporary solution
+static char *get_runpath() { return get_origin(); }
+#endif
+
+static char *get_first_runpath(const char *runpath) {
+    if (runpath == NULL)
+        return NULL;
+    _cleanup_free_ char *runpath1 = strdup(runpath);
+    if (runpath1 == NULL)
+        return NULL;
+    // runpath is not empty, so strtok shouldn't return NULL when first call.
+    return strdup(strtok(runpath1, ":"));
+}
+
+#if defined(__unix__) || defined(__unix)
+static bool init_dynamic_driver_list() {
+    _cleanup_free_ char *LPAC_DRIVER_HOME = get_first_runpath(get_runpath());
+    if (LPAC_DRIVER_HOME == NULL)
+        return false;
+    if (!strcmp(LPAC_DRIVER_HOME, "$ORIGIN")) {
+        free(LPAC_DRIVER_HOME);
+        LPAC_DRIVER_HOME = get_origin();
+        if (LPAC_DRIVER_HOME == NULL)
+            return false;
+    }
+    LPAC_DRIVER_HOME = realloc(LPAC_DRIVER_HOME, strlen(LPAC_DRIVER_HOME) + 8 + 1);
+    if (LPAC_DRIVER_HOME == NULL)
+        return false;
+    strcat(LPAC_DRIVER_HOME, "/drivers");
+
+    _cleanup_dir_ DIR *driver_dir = opendir(LPAC_DRIVER_HOME);
+    if (driver_dir == NULL) {
+        return true;
+    }
+#    if defined(_WIN32)
+    const char *suffix = ".dll";
+#    elif defined(__APPLE__) && defined(__MACH__)
+    const char *suffix = ".dylib";
+#    elif defined(__unix__) || defined(__unix)
+    const char *suffix = ".so";
+#    else
+#        error "Unsupported platform."
```

Comment:
may be needs update documenation, add supported platform list

Reply:
The platform list is hard to list... And I can't imagine which platform will reach this line, I leave it just for completeness...